tensorbay/cli/log.py

```diff
@@ -158,10 +164,179 @@ def _get_full_log(commit: Commit, branch_name: Optional[str]) -> str:
+    sorted_leaves: List[_CommitNode],
+    printer: Callable[[Commit, Optional[str], int, int, int], str],
+) -> Iterator[str]:
+    printing_amount = 1
+    printing_pointer = 0
+    while True:
+        old_print_pointer = printing_pointer
+        current_node = sorted_leaves[printing_pointer]
+        if current_node.available_child_num == 0:
+            parent = current_node.parent
+            if not parent:
+                break
+            if parent in sorted_leaves:
+                # Merge branch.
+                old_print_pointer = printing_pointer
+                printing_amount -= 1
+                printing_pointer = sorted_leaves.index(parent)
+                del sorted_leaves[old_print_pointer]
+            print_branch_name = ", ".join(
+                commit_id_to_branches.get(current_node.commit.commit_id, ())
+            )
+            yield printer(
+                current_node.commit,
+                print_branch_name,
+                printing_amount,
+                printing_pointer,
+                old_print_pointer,
+            )
+            parent.available_child_num -= 1
+            sorted_leaves[printing_pointer] = parent
+        else:
+            printing_amount += 1
+            printing_pointer += 1
+
+
+def _add_graph(
+    func: Callable[[Commit, Optional[str]], str]
+) -> Callable[[Commit, Optional[str], int, int, int], str]:
+    def wrapper(
+        commit: Commit,
+        branch_name: Optional[str],
+        printing_amount: int,
+        printing_pointer: int,
+        old_print_pointer: int,
+    ) -> str:
+        log = func(commit, branch_name)
+        split_lines = log.splitlines()
+        prefix = ["|"] * printing_amount
+        if printing_pointer == old_print_pointer:
+            details_prefix = " ".join(prefix)
+            details = "".join(
+                map(lambda item: f"{details_prefix} {item}\n", islice(split_lines, 1, None))
+            )
+            title_prefix = _get_title_prefix(old_print_pointer, prefix)
+            return f"{title_prefix} {split_lines[0]}\n{details}"
+        # Merge branches.
+        title_prefix = _get_title_prefix(old_print_pointer, prefix)
+        merge_prefix = _get_merge_graph(printing_amount, printing_pointer, old_print_pointer)
+        # Show one line commit.
+        if len(split_lines) == 1:
+            merge_prefix_graph = "\n".join(merge_prefix)
+            return f"{title_prefix} {split_lines[0]}\n{merge_prefix_graph}\n"
+        # Show complete commit.
+        details = "\n".join(_combine_details(merge_prefix, split_lines[1:]))
+        return f"{title_prefix} {split_lines[0]}\n{details}\n"
+
+    return wrapper
+
+
+def _get_merge_graph(
+    printing_amount: int, printing_pointer: int, old_print_pointer: int
+) -> List[str]:
+    graph_prefix = ["|", " "] * printing_amount
+    merge_prefix = list()
+    for index in range(old_print_pointer, printing_pointer, -1):
+        temp_prefix = graph_prefix[:]
```

Comment:
why do you need a copy of `graph_prefix`

Reply:
In every loop, I need original `graph_prefix`

Reply:
can we generate the `graph_prefix` in each loop?
`temp_prefix = ["|", " "] * printing_amount`

Reply:
OK～